search_params.go

```diff
@@ -0,0 +1,103 @@
+package faiss
+
+/*
+#include <faiss/c_api/Index_c.h>
+#include <faiss/c_api/IndexIVF_c.h>
+#include <faiss/c_api/impl/AuxIndexStructures_c.h>
+*/
+import "C"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+)
+
+type SearchParams struct {
+	sp *C.FaissSearchParameters
+}
+
+// Delete frees the memory associated with s.
+func (s *SearchParams) Delete() {
+	if s == nil || s.sp == nil {
+		return
+	}
+	C.faiss_SearchParameters_free(s.sp)
+}
+
+type searchParamsIVF struct {
+	NprobePct   float32 `json:"nprobe_pct,omitempty"`
+	MaxCodesPct float32 `json:"max_codes_pct,omitempty"`
+}
+
+func (s *searchParamsIVF) Validate() error {
+	if s.NprobePct < 0 || s.NprobePct > 100 {
+		return fmt.Errorf("invalid IVF search params, nprobe_pct:%v, should be in range [0, 100]", s.NprobePct)
+	}
+
+	if s.MaxCodesPct < 0 || s.MaxCodesPct > 100 {
+		return fmt.Errorf("invalid IVF search params, max_codes_pct:%v, should be in range [0, 100]", s.MaxCodesPct)
+	}
+
+	return nil
+}
+
+// Attempt to create IVF search params
+// and fall back to default search params if the index is not IVF.
+//
+// Always return a valid SearchParams object,
+// thus caller must clean up the object
+// by invoking Delete() method, even if an error is returned.
+func NewSearchParamsIVF(idx Index, params json.RawMessage, sel *C.FaissIDSelector,
+) (*SearchParams, error) {
+
+	rv := &SearchParams{}
+	if c := C.faiss_SearchParameters_new(&rv.sp, sel); c != 0 {
```

Comment:
Can avoid all these allocations is params is nil, same point as above.

Reply:
We can avoid this when both
` params==nil and len(exclude) == 0`

I have added this check.

```
  if params == nil and len(exclude) > 0 {
      // We can't avoid allocation of SearchParams in this case. 
      // SearchParams have 3 elements: Selector, Nprobe and Max_codes
      // In case user didn't supply ivf_nprobe_pct, we will still have to explicitly set Nprobe to 
      //  the value chosen at index creation time.
  }
  
```

Thanks